src/Field.php

```diff
@@ -204,96 +218,133 @@ public function __construct($defaults = [])
      */
     public function normalize($value)
     {
-        if (!$this->owner->strict_types) {
-            return $value;
-        }
-        if ($value === null) {
-            return;
-        }
-        $f = $this;
-
-        // only string type fields can use empty string as legit value, for all
-        // other field types empty value is the same as no-value, nothing or null
-        if ($f->type && $f->type != 'string' && $value === '') {
-            return;
-        }
-
-        switch ($f->type) {
-        case 'string':
-            if (!is_scalar($value)) {
-                throw new Exception('Field value must be a string');
-            }
-            $value = trim($value);
-            break;
-        case 'integer':
-            if (!is_numeric($value)) {
-                throw new Exception('Field value must be an integer');
+        try {
+            if (!$this->owner->strict_types) {
+                return $value;
             }
-            $value = (int) $value;
-            break;
-        case 'float':
-            if (!is_numeric($value)) {
-                throw new Exception('Field value must be a float');
+            if ($value === null) {
+                return;
             }
-            $value = (float) $value;
-            break;
-        case 'money':
-            if (!is_numeric($value)) {
-                throw new Exception('Field value must be numeric');
+            $f = $this;
+
+            // only string type fields can use empty string as legit value, for all
+            // other field types empty value is the same as no-value, nothing or null
+            if ($f->type && $f->type != 'string' && $value === '') {
+                if ($this->required && empty($value)) {
+                    throw new Exception('may not be a zero');
+                }
+
+                return;
             }
-            $value = round($value, 4);
-            break;
-        case 'boolean':
-            if (is_bool($value)) {
+
+            switch ($f->type) {
+            case null:
```

Comment:
Switch statement use loose comparison, so `null` can't be used here because it will be the same as `""` or `0`.
See also http://stackoverflow.com/a/7091795/1466341